File: sdk/php/src/ABAC.php
<?php
declare(strict_types=1);

namespace Ggid\Sdk;

/**
 * ABAC trait — provides attribute-based access control methods.
 *
 * Intended to be used by GGIDClient.
 */
trait ABAC
{
    /**
     * Evaluate ABAC policy with structured conditions.
     *
     * @param string $token Access token
     * @param string $action Action to evaluate (e.g. "transfer")
     * @param string $resource Resource being accessed (e.g. "inventory")
     * @param string $subject Subject identifier (user ID)
     * @param array $conditions Array of {field, operator, value} condition objects
     * @param string|null $tenantId Override tenant ID
     * @return ABACResult
     */
    public function evaluateABAC(
        string $token,
        string $action,
        string $resource,
        string $subject,
        array $conditions = [],
        ?string $tenantId = null,
    ): ABACResult {
        $body = [
            'action' => $action,
            'resource' => $resource,
            'subject' => $subject,
        ];
        if (!empty($conditions)) {
            $body['conditions'] = $conditions;
        }
        if ($tenantId !== null) {
            $body['tenant_id'] = $tenantId;
        }
        $data = $this->request('POST', '/api/v1/policies/abac/evaluate', $body, $token);
        return ABACResult::fromArray($data);
    }

    /**
     * Full ABAC policy check with subject context.
     *
     * @param string $token Access token
     * @param string $subject Subject identifier
     * @param string $resource Resource being accessed
     * @param string $action Action to evaluate
     * @param array $context Additional context attributes
     * @return ABACResult
     */
    public function checkPolicy(
        string $token,
        string $subject,
        string $resource,
        string $action,
        array $context = [],
    ): ABACResult {
        $body = [
            'subject' => $subject,
            'resource' => $resource,
            'action' => $action,
        ];
        if (!empty($context)) {
            $body['context'] = $context;
        }
        $data = $this->request('POST', '/api/v1/policies/abac/evaluate', $body, $token);
        return ABACResult::fromArray($data);
    }
}

File: sdk/rust/src/abac.rs
use serde_json::json;

use crate::client::GGIDClient;
use crate::error::GGIDError;
use crate::types::*;

pub struct ABACService {
    client: GGIDClient,
}

impl ABACService {
    pub(crate) fn new(client: GGIDClient) -> Self {
        Self { client }
    }

    /// Evaluate ABAC conditions against GGID policy engine.
    pub async fn evaluate_abac(
        &self,
        token: &str,
        request: ABACEvalRequest,
    ) -> Result<ABACEvalResult, GGIDError> {
        let resp = self
            .client
            .http
            .post(format!("{}/api/v1/policies/abac/evaluate", self.client.base_url))
            .header("Authorization", format!("Bearer {}", token))
            .header("X-Tenant-ID", &self.client.tenant_id)
            .json(&request)
            .send()
            .await?;

        if !resp.status().is_success() {
            let status = resp.status().as_u16();
            let body = resp.text().await.unwrap_or_default();
            return Err(GGIDError::Api { status, body });
        }

        Ok(resp.json().await?)
    }

    /// Check a policy with full subject and condition context.
    pub async fn check_policy(
        &self,
        token: &str,
        request: PolicyCheckRequest,
    ) -> Result<PolicyCheckResult, GGIDError> {
        let resp = self
            .client
            .http
            .post(format!("{}/api/v1/policies/check", self.client.base_url))
            .header("Authorization", format!("Bearer {}", token))
            .header("X-Tenant-ID", &self.client.tenant_id)
            .json(&request)
            .send()
            .await?;

        if !resp.status().is_success() {
            let status = resp.status().as_u16();
            let body = resp.text().await.unwrap_or_default();
            return Err(GGIDError::Api { status, body });
        }

        Ok(resp.json().await?)
    }
}
